src/chan_vese.jl

```diff
@@ -0,0 +1,252 @@
+using MosaicViews
+using LazyArrays
+using BenchmarkTools
+using Images, TestImages
+using ImageBase.ImageCore: GenericGrayImage, GenericImage
+
+function calculate_averages(img::AbstractArray{T, N}, H𝚽::AbstractArray{T, M}) where {T<:Real, N, M}
+    H𝚽ⁱ = @. 1. - H𝚽
+    ∫H𝚽 = sum(H𝚽)
+    ∫H𝚽ⁱ = sum(H𝚽ⁱ)
+    if ndims(img) == 2
+        ∫u₀H𝚽 = sum(img .* H𝚽)
+        ∫u₀H𝚽ⁱ = sum(img .* H𝚽ⁱ)
+    elseif ndims(img) == 3
+        ∫u₀H𝚽 = sum(img .* H𝚽, dims=(1, 2))
+        ∫u₀H𝚽ⁱ = sum(img .* H𝚽ⁱ, dims=(1, 2))
+    end
+    if ∫H𝚽 != 0
+        c₁ = ∫u₀H𝚽 / ∫H𝚽
+    end
+    if ∫H𝚽ⁱ != 0
+        c₂ = ∫u₀H𝚽ⁱ / ∫H𝚽ⁱ
+    end
+
+    return c₁, c₂
+end
+
+function difference_from_average_term(img::AbstractArray{T, N}, H𝚽::AbstractArray{T, M}, λ₁::Float64, λ₂::Float64) where {T<:Real, N, M}
+    c₁, c₂ = calculate_averages(img, H𝚽)
+
+    if ndims(img) == 2
+        return @. -λ₁ * (img - c₁)^2 + λ₂ * (img - c₂)^2
+    elseif ndims(img) == 3
+        return -λ₁ .* sum((img .- c₁).^2, dims=3) .+ λ₂ .* sum((img .- c₂).^2, dims=3)
+    end
+end
+# H𝚽 = LazyArray(@~ @. 1. * (𝚽ⁿ > 0))
+function _calculate_averages(img::AbstractArray{T, N}, 𝚽ⁿ::AbstractArray{T, M}) where {T<:Real, N, M}
+    ∫H𝚽 = ∫H𝚽ⁱ = ∫u₀H𝚽 = ∫u₀H𝚽ⁱ = 0
+
+    for i in CartesianIndices(img)
+        H𝚽 = 1. * (𝚽ⁿ[i] > 0)
+        H𝚽ⁱ = 1. - H𝚽
+        ∫H𝚽 += H𝚽
+        ∫H𝚽ⁱ += H𝚽ⁱ
+        ∫u₀H𝚽 += img[i] * H𝚽
+        ∫u₀H𝚽ⁱ += img[i] * H𝚽ⁱ
+    end
+    if ∫H𝚽 != 0
+        c₁ = ∫u₀H𝚽 ./ ∫H𝚽
+    end
+    if ∫H𝚽ⁱ != 0
+        c₂ = ∫u₀H𝚽ⁱ ./ ∫H𝚽ⁱ
+    end
+
+    return c₁, c₂
+end
+
+function δₕ(x::AbstractArray{T,N}, h::Float64=1.0) where {T<:Real, N}
+    return @~ @. h / (h^2 + x^2)
+end
+
+function initial_level_set(shape::Tuple)
+    x₀ = reshape(collect(0:shape[begin]-1), shape[begin], 1)
+    y₀ = reshape(collect(0:shape[begin+1]-1), 1, shape[begin+1])
+    𝚽₀ = @. sin(pi / 5 * x₀) * sin(pi / 5 * y₀)
+end
+
+function chan_vese(img::GenericGrayImage;
+                    μ::Float64=0.25,
+                    λ₁::Float64=1.0,
+                    λ₂::Float64=1.0,
+                    tol::Float64=1e-3,
+                    max_iter::Int64=500,
+                    Δt::Float64=0.5,
+                    reinitial_flag::Bool=false) #where {T<:Real, N}
+    img = float64.(channelview(img))
```

Comment:
Now I see where you are getting the 3d handling from: it's not actually 3d, it's your color channel. This is almost never needed in JuliaImages; just use the image as provided, and if you've written everything generically it should "just work." That is, for a grayscale image `c₁` will be a number or `Gray`, but for RGB `c₁` will be an RGB. And then you can truly support both 2d and 3d images by writing your code in a manner that generalizes across dimensions. So much easier than how other languages do these things, no?

Reply:
If you want a raw numerical array, then just:

```suggestion
    img = Float64.(img)
```

But generally, we encourage directly handling `Array{<:Colorant}` without `channelview` because it permits more generic implementation as Tim said in the previous comment.

Reply:
I would recommend deleting this line and the `minimum`/`maximum` standardization unless there is strong reason to keep it. Other languages may have done this just to commonize among `UInt8` images (ranging from 0:255), `UInt16` images (ranging from 0:65535), and `Float32`/`Float64` images (ranging from 0 to 1). But thanks to `N0f8` and `N0f16`, we've already done that. I think it's better to accept the image at face value.